src/syft/core/node/common/action/function_or_constructor_action.py

```diff
@@ -33,29 +41,144 @@ def __init__(
         #  id on the object directly
         self.id_at_location = id_at_location
 
+    @staticmethod
+    def intersect_keys(
+        left: Union[Dict[VerifyKey, UID], None], right: Dict[VerifyKey, UID]
+    ) -> Dict[VerifyKey, UID]:
+        # FIXME duplicated in run_class_method_action.py
+        # get the intersection of the dict keys, the value is the request_id
+        # if the request_id is different for some reason we still want to keep it,
+        # so only intersect the keys and then copy those over from the main dict
+        # into a new one
+        if left is None:
+            return right
+        intersection = set(left.keys()).intersection(right.keys())
+        # left and right have the same keys
+        return {k: left[k] for k in intersection}
+
     def execute_action(self, node: AbstractNode, verify_key: VerifyKey) -> None:
         method = node.lib_ast(self.path)
 
+        result_read_permissions = None
+
         resolved_args = list()
         for arg in self.args:
-            if isinstance(arg, Pointer):
-                # QUESTION: Where is get_object definied?
-                r_arg = node.store.get_object(id=arg.id_at_location)
-                resolved_args.append(r_arg)
-            else:
-                resolved_args.append(arg)
+            if not isinstance(arg, Pointer):
+                raise ValueError(
+                    f"args attribute of RunFunctionOrConstructorAction should only contain Pointers. "
+                    f"Got {arg} of type {type(arg)}"
+                )
+
+            r_arg = node.store.get_object(id=arg.id_at_location)
+            result_read_permissions = self.intersect_keys(
+                result_read_permissions, r_arg.read_permissions
+            )
+            resolved_args.append(r_arg.data)
 
         resolved_kwargs = {}
         for arg_name, arg in self.kwargs.items():
-            if isinstance(arg, Pointer):
-                r_arg = node.store.get_object(id=arg.id_at_location)
-                resolved_kwargs[arg_name] = r_arg
-            else:
-                resolved_kwargs[arg_name] = arg
+            if not isinstance(arg, Pointer):
+                raise ValueError(
+                    f"kwargs attribute of RunFunctionOrConstructorAction should only contain Pointers. "
+                    f"Got {arg} of type {type(arg)}"
+                )
+
+            r_arg = node.store.get_object(id=arg.id_at_location)
+            result_read_permissions = self.intersect_keys(
+                result_read_permissions, r_arg.read_permissions
+            )
+            resolved_kwargs[arg_name] = r_arg.data
 
         result = method(*resolved_args, **resolved_kwargs)
 
+        if isprimitive(value=result):
+            # Wrap in a PyPrimitive
+            result = PyPrimitive(data=result, id=self.id_at_location)
```

Comment:
Would we not have to wrap the actual primitive type instead of the parent class?

Reply:
Maybe it would be a good idea to add a convenience function to primitive types to make this easier? Or a utility function?

Reply:
CC: @tudorcebere

Reply:
Yep, I think this will change with Tudor's work!
This is based on Madhava's primitive type handling